web/src/lib/components/shared-components/leaflet/asset-marker-cluster.svelte

```diff
@@ -0,0 +1,90 @@
+<script lang="ts" context="module">
+	import { createContext } from '$lib/utils/context';
+	import L from 'leaflet';
+
+	const { get: getContext, set: setClusterContext } = createContext<() => L.MarkerClusterGroup>();
+
+	export const getClusterContext = () => {
+		return getContext()();
+	};
+</script>
+
+<script lang="ts">
+	import 'leaflet.markercluster';
+	import { onDestroy, onMount } from 'svelte';
+	import { getMapContext } from './map.svelte';
+	import { MapMarkerResponseDto, api } from '@api';
+	import { createEventDispatcher } from 'svelte';
+
+	class AssetMarker extends L.Marker {
+		marker: MapMarkerResponseDto;
+
+		constructor(marker: MapMarkerResponseDto) {
+			super([marker.lat, marker.lon], {
+				icon: new L.Icon({
+					iconUrl: api.getAssetThumbnailUrl(marker.id),
+					iconRetinaUrl: api.getAssetThumbnailUrl(marker.id),
+					iconSize: [60, 60],
+					iconAnchor: [12, 41],
+					popupAnchor: [1, -34],
+					tooltipAnchor: [16, -28],
+					shadowSize: [41, 41]
+				})
+			});
+
+			this.marker = marker;
+		}
+
+		getAssetId(): string {
+			return this.marker.id;
+		}
+	}
+
+	const dispatch = createEventDispatcher<{ view: { assets: string[] } }>();
+
+	export let markers: MapMarkerResponseDto[];
+
+	const map = getMapContext();
+
+	let cluster: L.MarkerClusterGroup;
+
+	setClusterContext(() => cluster);
+
+	onMount(() => {
+		cluster = new L.MarkerClusterGroup({
+			showCoverageOnHover: false,
+			zoomToBoundsOnClick: false,
+			spiderfyOnMaxZoom: false,
+			maxClusterRadius: 30,
+			spiderLegPolylineOptions: { opacity: 0 },
+			spiderfyDistanceMultiplier: 3
+		});
+
+		cluster.on('clusterclick', (event: L.LeafletEvent) => {
+			const ids = event.sourceTarget
+				.getAllChildMarkers()
+				.map((marker: AssetMarker) => marker.getAssetId());
+			dispatch('view', { assets: ids });
+		});
+
+		for (let marker of markers) {
+			const leafletMarker = new AssetMarker(marker);
+
+			leafletMarker.on('click', () => {
+				dispatch('view', { assets: [marker.id] });
+			});
+
+			cluster.addLayer(leafletMarker);
+		}
```

Comment:
How does this perform with a large amount of assets? Will this block the main thread for a long time?

Reply:
It's reasonably quick. I worked with the library before and even hundreds of thousands of markers are not an issue.

Reply:
I quickly checked on my prod instance. Absolutely no problem with 20k assets.

Reply:
Thanks for testing. After 10k assets I'm starting to see *some* slowdown, where loading takes ~500ms and seems to increase linearly. That's pretty good and definitely usable, but we should keep an eye on performance especially for users with a large library.